proposals/env.md

```diff
@@ -0,0 +1,591 @@
+reconstruct source-level location information (filename, line, and column) which
+enables (1). However, the source map format does not encode enough information
+for JavaScript debuggers to solve (2).
+
+This document describes a proposed extension to the source map format to enable
+JavaScript debuggers to provide (2) for languages that use JavaScript as a
+compilation target.
+
+### Goals
+
+1. Enabling JavaScript debuggers to rematerialize source-level scopes and
+   bindings when debugging JavaScript emitted by a compiler.
+
+2. Locating a source-level binding's JavaScript value. This encompasses
+   situations where:
+
+    * The binding was renamed in the compiled JS. For example, the binding is
+      named `atom?` in the original source and renamed to `atom_question` in the
+      compiled JavaScript. Or a minifier renames `getLatestFooWidget` to `aF`.
+
+    * The binding does not have a corresponding binding in the compiled
+      JavaScript. For example, the compiler inlined the binding's value after
+      recognizing that the binding was never mutated.
+
+3. Remain backwards compatible with the current source map revision 3 format.
+
+4. Remain extensible in such a way that future additions or modifications can
+   trivially remain backwards compatible with this proposal.
+
+5. Compact data representation.
+
+#### Non-Goals
+
+* Custom formatting of source-level values within JavaScript debuggers beyond
+  giving the debugger an arbitrary JavaScript value for a given binding. This is
+  outside of the scope of this proposal.
+
+## Syntax
+
+This proposal adds an `"env"` property to source maps, which contains a string
+of base 64 VLQ numbers. Legacy consumers and generators do not use the `"env"`
+property, thus goal (3) is satisfied.
+
+The basic syntactic unit is a "record" which has a type and a open ended set of
+property/value pairs. Records are loosely inspired by DWARF's Debugging
+Information Entries, but in a "source mappy" way. While the grammar below
+defines known record types, known properties, and known values, it explicitly
+allows for unknown record types, unknown properties, and unknown values for use
+by future extensions. Thus goal (4) is satisfied.
+
+In order to satisfy goal (5), we provide an abbreviation mechanism for records
+(again, similar to DWARF). This allows for the definition of an abbreviation
+that includes a record type and the set of properties. Then, when serializing a
+record whose type and property set already has an abbreviation definition, the
+record's type and properties can be omitted, emitting only the property
+values. Additionally, a property's value is encoded relative to the last value
+emitted for that property. This technique is used in encoding segments in the
+`"mappings"` property of the source map format, and has proved valuable in
+reducing file size.
+
+The `"env"` property can be parsed with the following BNF-like grammar, where
```

Comment:
Concatenation of source files, into final output files is extremely common. One proposed aspect of sourceMap, are sections. Those haven't been adopted, but if they would. It would simplified by the concatenation tooling implementations (and performance) and I believe provide similar benefit to debugging tools utilizing source-maps. Given then, should `env` also have `envSections` or something similar?